src/connection/adapter/base_http_adapter.ts

```diff
@@ -178,37 +181,41 @@ export abstract class BaseHttpAdapter implements Connection {
         removeRequestListeners()
       }
 
+      function onAbortSignal(): void {
+        abortController.abort('The request was aborted.')
+      }
+
       function removeRequestListeners(): void {
+        clearTimeout(timeout)
         request.removeListener('response', onResponse)
         request.removeListener('error', onError)
-        request.removeListener('timeout', onTimeout)
-        request.removeListener('abort', onAbort)
         request.removeListener('close', onClose)
         if (params.abort_signal !== undefined) {
-          if (isEventTarget(params.abort_signal)) {
-            params.abort_signal.removeEventListener('abort', onAbortSignal)
-          } else {
-            // @ts-expect-error if it's EventEmitter
-            params.abort_signal.removeListener('abort', onAbortSignal)
-          }
+          params.abort_signal.removeEventListener('abort', onAbortSignal)
         }
       }
 
-      if (params.abort_signal) {
-        // We should use signal API when nodejs v14 is not supported anymore.
-        // However, it seems that Http.request doesn't abort after 'response' event.
-        // Requires an additional investigation
-        // https://nodejs.org/api/globals.html#class-abortsignal
+      request.on('response', onResponse)
+      request.on('error', onError)
+      request.on('close', onClose)
+
+      if (params.abort_signal !== undefined) {
         params.abort_signal.addEventListener('abort', onAbortSignal, {
           once: true,
         })
       }
 
-      request.on('response', onResponse)
-      request.on('timeout', onTimeout)
-      request.on('error', onError)
-      request.on('abort', onAbort)
-      request.on('close', onClose)
+      abortController.signal.addEventListener(
+        'abort',
+        () => {
```

Comment:
How do we remove the subscription? Might it cause a memory leak?

Reply:
it's fixed already by https://github.com/ClickHouse/clickhouse-js/pull/167/commits/a796f8f6283d645ee367724cf4dee786eb449522